src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -345,13 +345,21 @@ IMergeTreeDataPart::~IMergeTreeDataPart()
 
 const IMergeTreeDataPart::Index & IMergeTreeDataPart::getIndex() const
 {
+    std::scoped_lock lock(index_mutex);
+    if (!index_loaded)
+        loadIndex(lock);
```

Comment:
I think that another potential problem is that all index will go to the user's memory usage. Maybe disable memory tracker for user while loading index like we do for loading marks?